packages/nx/src/hasher/task-hasher.ts

```diff
@@ -429,13 +429,14 @@ class TaskHasherImpl {
         partialHashes.push(node.data.version);
       }
       // we want to calculate the hash of the entire dependency tree
-      if (this.projectGraph.dependencies[projectName]) {
-        this.projectGraph.dependencies[projectName].forEach((d) => {
-          if (!visited.has(d.target)) {
-            partialHashes.push(this.hashExternalDependency(d.target, visited));
-          }
-        });
-      }
+      this.projectGraph.dependencies[projectName]?.forEach((d) => {
+        if (!visited.has(d.target)) {
+          partialHashes.push(
+            this.hashExternalDependency(d.target, new Set(visited))
```

Comment:
I don't like that a new `Set` is created each time. A lot more memory will be used.

When we hash an external dependency, there is always a `source` and a `target`.

As you noticed ,when `A -> C` and `B -> C`, these are two different dependencies which should both be visited. I think the `visited` Set should have both `source` and `target` as items. When C is hashed for A, it should record that `A -> C` has been visited. and when C is then visited to be hashed for `B -> C` it won't find it in the `Set` and it will hash.

So I think the signature here should change to:

```ts
  private hashExternalDependency(
    sourceProjectName: string,
    targetProjectName: string,
    visited = new Set<string>()
  ): string {
    visited.add(`${sourceProjectName}->${targetProjectName}`);
    ...
```

Reply:
I have changed the code accordingly, let me know if it is correct.